### ISSUE TITLE
Add push options to the mike deploy command

### ISSUE THREAD

**Issue by gusmb** [jimporter/mike#118]
It would be nice to specify push options when pushing to the remote branch while deploying. A use case is to use
the "ci.skip" option to instruct Gitlab to skip the CI for the docs branch pushed by the mike deploy command....

https://docs.gitlab.com/ee/user/project/push_options.html

In my case I always see an ugly failed pipeline as non-existing for the docs branch since our Gitlab Server enforces the run of a pipeline.
There might be other use cases for push options.

**Comment by jimporter** [comment]
The `-p`/`--push` option is just a shorthand for the common case of running `git push origin gh-pages`. I think all you should need to do is:
```
mike deploy
git push origin gh-pages --push-option=HELLO
```

**Comment by gusmb** [comment]
That works as well, I was just hoping that this could be handled natively with the mike CLI when using that shorthand as it is a common thing, instead of having to use the git CLI. I tried and it works but seems a bit redundant as I know need to give the remote to both mike and GIT. Pushing to origin fails from the pipeline as I need to set the correct access token

**Comment by jimporter** [comment]
I'd prefer to avoid adding more options to mike unless absolutely necessary. When it comes to doing non-default things with Git, I think it would be best to write the exact Git command so that it's maximally explicit. That way there's no confusion about what's really getting executed.

I guess I could see adding an option that forwards *arbitrary* args to subcommands, so that you could do `mike deploy -X,--push-option=HELLO`. I'm not sure though, since I think the explicit Git command makes it more obvious what's happening.

**Comment by gusmb** [comment]
Fair enough. I guess in that case documentation examples around use cases like the push option to skip the CI etc, would be good so it's clear when to skip the -p option and handle the git separately with git CLI. The example you gave me earlier was helpful and worked for me with minor tweaks.

**Comment by jimporter** [comment]
One thing I've been considering here is that it's probably worth adding support for some set of environment variables to control Git behavior. For example, a user might want to use a specific version of Git not in their path, and so would set `GIT=/path/to/git`. Similarly, we could support this issue via a `GIT_PUSH_FLAGS` variable. Then there would be similar env vars for each other Git subcommand we use. I'm not sure how useful that is, but it would be the most general-purpose solution to the problem...

**Comment by gusmb** [comment]
Yeah seems like adding support for git options so that the user can consolidate the docs version management with just the mike CLI instead of part with mike and part with git would help keep things more tidy

**Comment by jakob1379** [comment]
Mike v. 2.1.3 can do this, I guess its safe to close this?

```bash
❯ mike deploy --help
usage: mike deploy [-h] [-t TITLE] [-u] [--alias-type TYPE] [-T TEMPLATE] [-F FILE] [-r REMOTE] [-b BRANCH] [-m MESSAGE] [-p] [--allow-empty] [--deploy-prefix PATH]
                   [--ignore-remote-status] [--prop-set PROP=JSON] [--prop-set-string PROP=STRING] [--prop-set-all JSON] [--prop-delete PROP] [--prop-delete-all]
                   VERSION [ALIAS ...]

Build the current documentation and deploy it to the specified version (and aliases, if any) on the target branch.

positional arguments:
  VERSION               version to deploy this build to
  ALIAS                 additional alias for this build

options:
  -h, --help            show this help message and exit
  -t, --title TITLE     short descriptive title for this version
  -u, --update-aliases  update aliases pointing to other versions
  --alias-type TYPE     method for creating aliases (one of: symlink, copy, redirect; default: symlink)
  -T, --template TEMPLATE
                        template file to use for redirects
  -F, --config-file FILE
                        the MkDocs configuration file to use

git arguments:
  -r, --remote REMOTE   origin to push to (default: origin)
  -b, --branch BRANCH   branch to commit to (default: gh-pages)
  -m, --message MESSAGE
                        commit message
  -p, --push            push to {remote}/{branch} after commit
  --allow-empty         allow commits with no changes
  --deploy-prefix PATH  subdirectory within {branch} where generated docs should be deployed to
  --ignore-remote-status
                        don't check status of remote branch

property manipulation arguments:
  --prop-set PROP=JSON  set the property at PROP to a JSON value
  --prop-set-string PROP=STRING
                        set the property at PROP to a STRING value
  --prop-set-all JSON   set all properties to a JSON value
  --prop-delete PROP    delete the property at PROP
  --prop-delete-all     delete all properties

❯ mike --version
mike 2.1.3
```

**Comment by jimporter** [comment]
> Mike v. 2.1.3 can do this, I guess its safe to close this?

Mike has had an option to push since it was first released, but that's different from this bug report, which is about passing additional options to `git push`. I'm still not sure if I'd want to add in more complexity here when using `git push` directly is an option though...

**Comment by jakob1379** [comment]
I misunderstood, then. Sorry!

I would argue that if you want to make more advanced commands it should be outside the scope of `mike` as to not make it a proxy for all git commands.
